Updated fip file standard readme to reflect actual patch cord situation

dougollerenshaw opened the pull request:
This PR clarifies terminology in the FIP standard and describes the current limitation that FIP acquisition metadata lacks a mechanism for documenting the patch cord to implanted fiber mapping in a given experiment. Instead, there is an assumption that the standard convention of `patch_cord_n` mapping to `fiber_n` is always followed.

This PR also standardizes notation by using lowercase "n" to refer to an individual index, uppercase "N" to represent the total number of implanted fibers (N ≤ 4 in our case) and uppercase "K" to represent the total number of patch_cords and associated ROIs (K = 4 in our case).

Closes #58

dougollerenshaw committed:
Updated fip file standard readme to reflect actual patch cord situation

dougollerenshaw committed:
Cleaned up notation

dougollerenshaw committed:
Generalized fip.md so that it doesn't assume a fixed number of patch cords/ROIs. Came from feedback from Kenta. Use variable  instead of 4

saskiad reviewed (commented):
see comments

dougollerenshaw committed:
stereotactic not stereotaxic

dougollerenshaw commented on docs/file_formats/fip.md:
```diff
@@ -8,6 +8,27 @@
+
+This document uses specific terminology to distinguish between different components of the FIP system:
+
+* **Fibers**: Optical fibers surgically implanted in the mouse's brain. A mouse may have 1-4 implanted fibers. Fiber implantation details (locations, stereotaxic coordinates) are documented in the procedures.json metadata file.
```
Thanks. Fixed. https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/4dad44b42f9a3392b7e3191083a5ef0c6ec30736

dougollerenshaw committed:
Added detail in Relationship to aind-data-schema section

dougollerenshaw commented on docs/file_formats/fip.md:
```diff
@@ -180,7 +203,13 @@ Data acquisition code that generates data in this format is available from the [
+
+* **procedures.json**: Documents implanted fiber locations (stereotaxic coordinates) and viral injection metadata (stereotaxic coordinates, materials, viral constructs)
+
+* **acquisition.json**: Documents which patch cords were connected to which implanted fibers during the experiment. This mapping is critical for correctly interpreting the data, as the `Fiber_n` columns in the CSV files represent patch_cord_n data, and experimenters may not always connect patch_cord_n to fiber_n (e.g., if a particular implanted fiber is damaged or if a different configuration is  ...
```
How's this? https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/7175a200c7737f38f56a93951b48d7d79d761cf4

dougollerenshaw committed:
Update to index notation

dougollerenshaw commented on docs/file_formats/fip.md:
```diff
@@ -8,6 +8,27 @@
+**Notation**: Throughout this document, we use:
+- **N** to represent the total number of implanted fibers in a given mouse (1 ≤ N ≤ K)
+- **K** to represent the total number of patch cords/ROIs defined by the hardware (current rigs use K = 4, but this document is written to allow other values)
+- **n** to represent a specific index (0-based, e.g., n ∈ {0, …, K-1})
```
Good point. I was being sloppy with notation. See here:
https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/ad62834365ec9aee296ff42d1a75bf25a378e535
I clarified so that:
* N is the total number of implanted fibers in the brain. n is the fiber index.
* K is the total number of patch cords/ROIs in the FIP system. k is the patch cord/ROI index.

So for now we assume that patch_cord_k connects to fiber_n, where k=n.

saskiad commented on docs/file_formats/fip.md:
```diff
@@ -180,7 +204,13 @@ Data acquisition code that generates data in this format is available from the [
+* **procedures.json**: Documents implanted fiber locations (stereotactic coordinates) and viral injection metadata (stereotactic coordinates, materials, viral constructs)
+
+* **acquisition.json**: Documents which patch cords were connected to which implanted fibers during the experiment. This is contained in the [connections](https://aind-data-schema.readthedocs.io/en/latest/components/connections.html#connection) list in the [DataStream object](https://aind-data-schema.readthedocs.io/en/latest/acquisition.html#datastream). This mapping is critical for correctly int ...
+
```
the link after [channel] looks like a copy/paste mistake

dougollerenshaw replied:
Good catch! Fixed in https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/818070b88477215aaa33cd387464e775ea05f610

saskiad reviewed (changes requested):
one small link fix. Otherwise looks good.

bruno-f-cruz commented on docs/file_formats/fip.md:
```diff
@@ -8,6 +8,28 @@
+
+* **Patch cords**: Fiber optic cables that are permanently part of the FIP hardware system. The current FIP system has exactly 4 patch cords. At experiment time, experimenters physically connect patch cords to implanted fibers. Not all patch cords are necessarily connected during every experiment.
+
+* **Fiber bundle**: The assembly of all patch cords whose termini are visible to the imaging cameras. The fiber bundle remains stationary during experiments.
```
I dont think the file format should specify this. Nothing is preventing the file format from accommodating multiple bundles per acquisition OR even moving bundles. In fact, we have previously floated the idea of connecting a rotary joint directly to the focusing element for freely moving experiments.

bruno-f-cruz commented on docs/file_formats/fip.md:
```diff
@@ -53,9 +75,11 @@ Each fiber photometry session will primarily be analyzed by using the average si
-* `Fiber_N` Average signal values for Fiber_N's selected ROI.
+* `Fiber_0`, `Fiber_1`, …, `Fiber_{K-1}` Average signal values from patch_cord_k's ROI (one column per patch cord/ROI)
+
+Note: There are always exactly K `Fiber_n` columns (n = 0 … K-1) corresponding to the configured patch cords and ROIs for the rig. If N < K (where N is the number of implanted fibers), then patch cords N through K-1 are unused and their corresponding columns contain data from unconnected patch cords.
```
As I mentioned in #58, if anything we should be simplifying this by removing the notion of Fiber and using ROI instead. I kept it for backwards compatibility with Kenta's CSV. Otherwise, given what I mentioned above, this description just further complicates the life of anyone trying to understand the file standard and respective implementation.

bruno-f-cruz reviewed (changes requested):
Thanks for submitting a proposal.

Briefly, I think that while the proposal definitely has documentation value, it is largely not related to file standards. Most of it pertains to concepts that are part of the aind-data-schema, and should thus be documented there (maybe in the `procedures` model or in the `read-the-docs`?).

The scope of this repository is to define file formats that are intentionally agnostic to many of the SOPs of particular systems. We already have aind-data-schema to track that sort of metadata, no need for duplication. Preserving this level of abstraction is important if the standard is to remain broadly applicable and maintainable over time. For guidance on distinguishing between a new standard and supporting documentation, please see: (please see https://github.com/AllenNeuralDynamics/aind-file-standards/blob/main/docs/core/contributing.md#is-it-a-new-standard-or-just-documentation for further information)

Finally, please see https://github.com/AllenNeuralDynamics/aind-file-standards/blob/main/docs/core/contributing.md#how-to-contribute that details how contributions should be made to this repository. These guidelines were reviewed and approved by contributors from multiple groups to ensure consistency and shared understanding. At this stage, it seems that there has not yet been sufficient discussion around motivation, potential impact and alternatives regarding the original issue (#58).

dougollerenshaw committed:
Fixed copy/paste error in channel link

dougollerenshaw committed:
generalize fiber number statement

dougollerenshaw committed:
Note file standard is agnostic to patch cord count

dougollerenshaw committed:
Removed line about stationary fiber bundle

dougollerenshaw committed:
Remove ROI count specification and clarify purpose

dougollerenshaw commented on docs/file_formats/fip.md:
```diff
@@ -193,6 +223,6 @@ The following are expected to be true for all FIP data collected under this stan
 * All rows in the `<color>.csv` files will be present in the corresponding camera metadata files. The opposite is not guaranteed to be true.
-* A `<color>.csv` file is not guaranteed to have a `Fiber_N` column. A `Background` column is always present. The order of the columns in the `<color>.csv` files is not guaranteed to be the same across different sessions. It is thus recommended to use the header as the index for the columns.
-* The naming of `Fiber_<i>` columns in the `<color>.csv` files is guaranteed to be sequential, starting from `Fiber_0` and going up to `Fiber_N`.
+* A `<color>.csv` file always contains exactly K `Fiber_n` columns (`Fiber_0` through `Fiber_{K-1}`) corresponding to the patch cords for the rig. A `Background` column is always present. The order of the columns in the `<color>.csv` files is not guaranteed to be the same across different sessions. It is thus recommended to use the header as the index for the columns.
```
Agreed - we shouldn't save it. But until https://github.com/AllenNeuralDynamics/Aind.Physiology.Fip/issues/40 is resolved, the system does produce all columns regardless of connected fibers. Users need to understand this when interpreting current files.

dyf commented:
I agree with @bruno-f-cruz's sentiment that some of this not really about the file standard. 

I also think that AIND-specific conventions like this need to be documented, and it's nice to have it close to this file.

What do folks think about creating a `docs/conventions/aind/fip.md` file, or something like that?

dougollerenshaw commented:
> I agree with @bruno-f-cruz's sentiment that some of this not really about the file standard.
> 
> I also think that AIND-specific conventions like this need to be documented, and it's nice to have it close to this file.
> 
> What do folks think about creating a `docs/conventions/aind/fip.md` file, or something like that?

That makes sense. I worry about splitting information across files, though. It just makes it hard to find things.

What about adding a section titled something like 'AIND-specific implementation details' that contains the information about the patch_cord to fiber mapping conventions/assumptions?

dougollerenshaw commented:
> Why not do what I initially suggested and store it with aind-data-schema documentation?

This information is vital for interpreting the data files themselves.

saskiad commented:
the data schema read the docs is definitely not the right place for it. The issue here isn't about documenting what the metadata fields are. It's primarily around how to understand the data in the files given how they are using their rigs - and what metadata fields help them interpret it. But the primarily issue is to articulate the assumptions that have to be true for it all to be interpretable.

bruno-f-cruz commented:
Sorry for the confusion, but could someone clearly articulate what aind-data-schema’s role is with respect to data analysis?

My understanding has been informed by the following description in the file standards repository:
https://github.com/AllenNeuralDynamics/aind-file-standards/blob/main/docs/core/template.md#relationship-to-aind-data-schema

> This section is reserved to describe how the data format relates and/or is represented by the aind-data-schema library. Examples include how the hardware and software metadata information should be encoded in the schemas, how data acquired should exist as Epochs and other critical conventions. It is important to note that this section should create a one-way dependency, where data formats should NOT depend on the schema and instead stand on their own. aind-data-schema should instead be used to provide extra information, context and validation to the data asset.

This seems consistent with the description in the aind-data-schema documentation:
https://aind-data-schema.readthedocs.io/en/latest/

> Data acquired at the Allen Institute for Neural Dynamics (AIND) is accompanied by metadata describing how it was acquired, processed, and analyzed. This metadata is stored in JSON files according to the schema defined in this library. Our goal in capturing this metadata is to make our data findable and understandable.

>The issue here isn't about documenting what the metadata fields are. It's primarily around how to understand the data in the files given how they are using their rigs - and what metadata fields help them interpret it. But the primarily issue is to articulate the assumptions that have to be true for it all to be interpretable.

I agree that this is "an" important issue. However, a file format standard alone is not trying to specify everything required to interpret an experiment. In practice, one would need substantially more context—for example: laser powers, laser types, injected viruses, behavioral task, etc. That type of information sounds like precisely what aind-data-schema is intended to capture.

If I understand correctly, the primary “job” of aind-data-schema is to provide the metadata and context necessary to interpret the data. However, it sounds like there may be a suggestion that this information should instead be encoded directly in the file format standards. This raises several edge cases for me:

* Does this mean that camera metadata (e.g., camera model, sensor size, etc.) should be part of the behavior video file format standard as well?
https://allenneuraldynamics.github.io/aind-file-standards/file_formats/behavior_videos

* Should surgery coordinates, target brain area, and similar information be included in the electrophysiology standard?
https://github.com/AllenNeuralDynamics/aind-file-standards/blob/main/docs/file_formats/ecephys.md

* Should the planar ophys (2p) standard include information about the brain coordinates being imaged?
https://github.com/AllenNeuralDynamics/aind-file-standards/blob/main/docs/file_formats/planar_ophys.md

I am happy to stop belaboring this discussion if people feel strongly about a particular direction and are willing to maintain it—especially if, as @dyf suggested, this work happens in a separate folder within the repository. However, for my own sanity, I would really appreciate a clearer articulation of how people currently view the roles and boundaries between file format standards, metadata and aind-data-schema

Understanding this division of responsibility would help me (and likely others) reason about where particular requirements should live going forward.

Thanks!

dougollerenshaw committed:
Clarify that this document does not describe schema

dougollerenshaw committed:
Remove discussion of patch cord to fiber from terminology section

dougollerenshaw committed:
Bump version from 0.4.0 to 0.5.0

dougollerenshaw commented:
The changes here are not meant to be AIND data schema specific. The goal was only to clarify some terminology to avoid more confusion going forward about what is contained in these files. There has been real confusion to date, both for others and myself (e.g., see the multi-month thread here: https://github.com/AllenNeuralDynamics/aind-metadata-mapper/issues/277), so my only goal here is to be clearer and more explicit to save time/confusion going forward.

But to help address some of Bruno's concerns, I made the following additional changes.

I made a couple more changes to generalize the text:
* Modified the "Relationship to aind-data-schema" section to make it clear that the file standard does not specify metadata schema, but that additional metadata is necessary to fully interpret the signals contained in the file. Linked to the readthedocs for the AIND data schema. See https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/3e3494034cb6b1053bf2b4ee84bb5d34bc589808
* I removed the discussion of the mapping between patch cords and fibers from the terminology section since this is already described in the "Relationship to aind-data-schema" section. See https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/80848d89080c37081dafdd7b695d09723fc516b3
* I bumped the version from "0.4.0" to "0.5.0". I think we could also make the argument that these changes only represent "bug fix" level changes since they don't actually change the standard in any way, but only clarify some details. In that case, maybe the version should actually be "0.4.1"? 

If there are additional concerns about specific lines of text, please call them out. But I think we should identify a pragmatic path to merging this PR and moving on.

saskiad commented on docs/file_formats/fip.md:
```diff
@@ -53,9 +75,11 @@ Each fiber photometry session will primarily be analyzed by using the average si
-* `Fiber_N` Average signal values for Fiber_N's selected ROI.
+* `Fiber_0`, `Fiber_1`, …, `Fiber_{K-1}` Average signal values from patch_cord_k's ROI (one column per patch cord/ROI)
+
+Note: There are always exactly K `Fiber_n` columns (n = 0 … K-1) corresponding to the configured patch cords and ROIs for the rig. If N < K (where N is the number of implanted fibers), then patch cords N through K-1 are unused and their corresponding columns contain data from unconnected patch cords.
```
the use of "n" in "Fiber_n" here is confusing. You are not referring to the "n" that is the patch cord index, but rather that it is an index, right? Maybe "i" instead?

saskiad replied:
it's possible we don't need to define "n" and "k" as patch cord and fiber indices above? that might help?

dougollerenshaw replied:
I dropped lowercase notation everywhere for simplicity. See https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/c927b8707552f7b1e72270b7618d38637c3a0fb8

saskiad commented on docs/file_formats/fip.md:
```diff
@@ -53,9 +75,11 @@ Each fiber photometry session will primarily be analyzed by using the average si
+
+Note: There are always exactly K `Fiber_n` columns (n = 0 … K-1) corresponding to the configured patch cords and ROIs for the rig. If N < K (where N is the number of implanted fibers), then patch cords N through K-1 are unused and their corresponding columns contain data from unconnected patch cords.
+
+**Note on naming convention**: The column names `Fiber_0`, `Fiber_1`, …, `Fiber_{K-1}` are somewhat misleading - these columns actually represent data from patch_cord_0, patch_cord_1, …, patch_cord_{K-1}, respectively. The naming persists for backward compatibility. The relationship between patch cords and implanted fibers in the mouse is not captured in this file format and should be documented  ...
```
The last sentence here is redundant with the section above.

dougollerenshaw replied:
Redundant sentence removed in https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/b18b53eb767f130c34a68d02c1d3ea4dfbc3240d

saskiad commented on docs/file_formats/fip.md:
```diff
@@ -180,7 +204,13 @@ Data acquisition code that generates data in this format is available from the [
+
+* **procedures.json**: Documents implanted fiber locations (stereotactic coordinates) and viral injection metadata (stereotactic coordinates, materials, viral constructs)
+
+* **acquisition.json**: Documents which patch cords were connected to which implanted fibers during the experiment. This is contained in the [connections](https://aind-data-schema.readthedocs.io/en/latest/components/connections.html#connection) list in the [DataStream object](https://aind-data-schema.readthedocs.io/en/latest/acquisition.html#datastream). This mapping is critical for correctly int ...
```
Last sentence:
- add "This" at the beginning to make it a complete sentence.
- put 'config' inside the bracket with channel, so [channel config] 
- last clause  -> "as provided by the scientist in the surgical request" instead of "assuming that ..."

dougollerenshaw replied:
Addressed here: https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/commits/916fe3297f9277633c6afd22fea8c3c72ca8284f

saskiad commented on docs/file_formats/fip.md:
```diff
@@ -180,7 +204,13 @@ Data acquisition code that generates data in this format is available from the [
+
+* **acquisition.json**: Documents which patch cords were connected to which implanted fibers during the experiment. This is contained in the [connections](https://aind-data-schema.readthedocs.io/en/latest/components/connections.html#connection) list in the [DataStream object](https://aind-data-schema.readthedocs.io/en/latest/acquisition.html#datastream). This mapping is critical for correctly int ...
+
+A standard convention is for patch cords 0 through N-1 to connect to fibers 0 through N-1, with patch cords N through K left unused, but this convention is not enforced by the acquisition software. There is currently no method for capturing intentional or inadvertent variances from this convention (again, see [this issue](https://github.com/AllenNeuralDynamics/Aind.Physiology.Fip/issues/40).)
```
the method for capturing this is to make sure the metadata is correct, no?

saskiad replied:
you mean there isn't a convenient/automated way to do this currently?

dougollerenshaw replied:
I mean there's no way to do this now, other than assuming the convention is followed.

Currently the extracted metadata generated by the fip acquisition code does not tell us anything about the patch cable to implanted fiber mapping. That's what's requested in issue https://github.com/AllenNeuralDynamics/Aind.Physiology.Fip/issues/40

In the FIP metadata mapper, we've hard-coded the assumption that patch_cord_0 is connected to fiber_0, patch_cord_1 is connected to fiber_1, etc. If there are less than 4 implanted fibers, we assume that the corresponding patch cables are unplugged. This issue documents the hard-coded assumption: https://github.com/AllenNeuralDynamics/aind-metadata-mapper/issues/416

All that said, we could potentially just remove the paragraph starting with "A standard convention...". As long as everyone is following the convention for now, the metadata will be correct. And after those issues are resolved, we'll have a mechanism for documenting the actual patch_cord to fiber mapping in the metadata, as described in the previous paragraph, if people do deviate from the convention.

saskiad replied:
But there is a field in the metadata for this information to be provided. So the way to do this is, if the convention is not followed, the person doing the experiment must manually edit the metadata. I think this is important to call out, because we need to be explicit about the assumption for this convention.

dougollerenshaw replied:
See changes here: https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59/changes/228eb544fe0d554959943e978dbb442cda69b1b1. This makes it explicit that scientists must manually update the metadata if the convention isn't followed until there's some ability to specify these connections at the rig.

saskiad reviewed (commented):
see comments

dougollerenshaw committed:
Merge branch 'main' into fix_58_improve_file_standard_doc

dougollerenshaw committed:
Simplify indexing text

dougollerenshaw committed:
Removed redundant sentence

dougollerenshaw committed:
Updated wording

dougollerenshaw committed:
Make explicit that scientists must update metadata if convention not followed

bruno-f-cruz reviewed (commented):
I still don't agree with the amount of coupling these changes introduce to the file format. By merging this, we are effectively locking ourselves in a very specific use of a file format that could, in theory, support other uses (e.g., HSFP, widefield, etc.). 

If I were to change anything about the standard, I would go in the opposite direction and rename the columns from "Fiber_"  to "ROI_" to prevent any further confusion (this was a choice from something Kenta was previously using), but leave it at that. 

I think this PR goes directly against some of the [core principles of this repository](https://github.com/AllenNeuralDynamics/aind-file-standards/blob/main/docs/core/contributing.md), and we should, as I mentioned, simply be systematic and decide once and for all that this is not about a standard, and instead it is about documentation (which is fine, but at least let's be clear about it). 

If this is what you have in mind, I would ask you to make a separate PR modifying the language in `./docs/core/*` files before merging this. 

I think you have enough approvals anyway, so I am def not blocking this.

---

For more context, see my previous concerns:

https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59#pullrequestreview-3590115540
https://github.com/AllenNeuralDynamics/aind-file-standards/issues/58#issuecomment-3609579880
https://github.com/AllenNeuralDynamics/aind-file-standards/pull/59#issuecomment-3676161840
